internal/stacks/stackmigrate/resources.go

```diff
@@ -0,0 +1,373 @@
+		// for the resource.
+		// If this is successful, we can now start adding source information
+		// to diagnostics.
+		diags = m.loadConfig(target)
+		if diags.HasErrors() {
+			m.emitDiags(diags)
+			continue
+		}
+		component := target.AbsResource.Component
+		componentAddr := target.AbsResource.Item
+
+		trackComponent(target)
+
+		// retrieve the provider that was uses to create the resource instance.
+		providerAddr, provider, diags := m.getOwningProvider(target)
+		if diags.HasErrors() {
+			m.emitDiags(diags)
+			continue
+		}
+
+		schema, _ := provider.GetProviderSchema().SchemaForResourceType(resource.Addr.Resource.Mode, resource.Addr.Resource.Type)
+		if schema == nil {
+			m.emitDiags(diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Resource type not found",
+				Detail:   fmt.Sprintf("Resource type %s not found in provider schema.", resource.Addr.Resource.Type),
+				Subject:  target.ModuleConfig.SourceAddrRange.Ptr(),
+			}))
+			continue
+		}
+
+		for instanceKey, instance := range resource.Instances {
+			instanceAddr := stackaddrs.AbsResourceInstance{
+				Component: component,
+				Item:      componentAddr.Instance(instanceKey),
+			}
+
+			m.emit(&stackstate.AppliedChangeResourceInstanceObject{
+				ResourceInstanceObjectAddr: stackaddrs.AbsResourceInstanceObject{
+					Component: instanceAddr.Component,
+					Item:      instanceAddr.Item.DeposedObject(addrs.NotDeposed),
+				},
+				NewStateSrc:        instance.Current,
+				ProviderConfigAddr: providerAddr,
+				Schema:             schema,
+			})
+
+			for deposedKey, deposed := range instance.Deposed {
+				m.emit(&stackstate.AppliedChangeResourceInstanceObject{
+					ResourceInstanceObjectAddr: stackaddrs.AbsResourceInstanceObject{
+						Component: instanceAddr.Component,
+						Item:      instanceAddr.Item.DeposedObject(deposedKey),
+					},
+					NewStateSrc:        deposed,
+					ProviderConfigAddr: providerAddr,
+					Schema:             schema,
+				})
+			}
+		}
+	}
+	return components
+}
+
+// search searches for the state resource in the resource mappings and when found, converts and returns the relevant
+// stackResource.
+//
+// If the resource or module is nested within the root module, they will be migrated to the component with the address structure retained.
+// For example, a resource with the address module.my_module.module.child.aws_instance.foo will be migrated to
+// component.my_component.module.child.aws_instance.foo if the corresponding map key is found.
+// E.g module.child.aws_instance.foo will be replaced with component.child.aws_instance.foo
+func (m *migration) search(resource addrs.AbsResource, resources map[string]string, modules map[string]string) (*stackResource, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+	ret := &stackResource{
+		ContainingModule: resource.Module.Module(),
+	}
+
+	parseComponentInstance := func(target string) (Instance, tfdiags.Diagnostics) {
```

Comment:
I think there are a couple of parsers under `internal/stacks/stackaddrs/components.go` that could be used here instead.

Reply:
Well, partially. This function is a wrapper around stackaddrs.ParseAbsComponentInstanceStrOnly, but we want to support bare component names, i.e "self" instead of "component.self"